rcl_yaml_param_parser/src/parser.c

```diff
@@ -38,6 +38,7 @@ typedef enum yaml_map_lvl_e
   MAP_UNINIT_LVL = 0U,
   MAP_NODE_NAME_LVL = 1U,
   MAP_PARAMS_LVL = 2U,
+  MAP_PARAMS_DESCRIPTORS_LVL = 3U
```

Comment:
@bpwilcox hmm, putting parameter descriptors within the `ros__parameters` and relying on users not using the `parameter__descriptors` key for their own purposes under the assumption that it's unlikely enough is not ideal. Consider either making this map a sibling of `ros__parameters` or supporting further nesting e.g.:

```yaml
my_node:
    ros__parameters:
        param1: 1.0 
        param2:
            value: foo
            descriptor:
                read_only: true
```

Reply:
I would think the further nesting could be more confusing to the user, just my opinion (it seems to conflate the yaml syntax rules for when it is a parameter vs a descriptor). I mentioned this in a another comment, but yes I can see `parameter__descriptors` as a sibling to `ros__parameters`,but as I did attempt that first, I think it is a bit more messy to write the code for doing so than simply using the key to trigger a new map level. If per your suggestion, we change the key to `ros__parameter_descriptors`, do you think it is unique enough for that assumption?

Reply:
I'm wary about the word `enough` there. I'd rather just preclude name collisions. But maybe others don't think alike. @ivanpauno @jacobperron @wjwwood thoughts?

Reply:
I don't see a reason to nest `parameter_descriptors` under `ros__parameters`, and I would rather implement it as a sibling of `ros__parameters`, as @hidmic suggested. `ros__parameter_descriptors` is a good name for that sibling.
I also agree that the alternative of further nesting is confusing.

Reply:
I prefer the extra nesting since it keeps the descriptors physically close to the parameters they are describing. As a sibling to `ros__parameters`, I think it requires more mental effort. My two cents.

Reply:
Well, actually that option isn't bad either. But both the value and the descriptor should be optional (and at least one has to be specified).

Reply:
I don't have a strong opinion. I see merit in both keeping it separate and having descriptors near the overrides.

Reply:
In regards to the extra nesting, I am unsure what would be a preferable syntax. The one @hidmic  presented as an example implicates that there are two different ways to define an override, the original way, and the nested version, where the descriptor is attached and there's 2 new key words, `value` and `descriptor`. In my opinion, I think it obfuscates the parameter namespacing syntax by adding even more key words to check.  For that reason, I'm still in favor of a separate map level (either sibling or scoped under `ros__parameters`) for the `ros__parameter_descriptors` where the rules for parsing are cleanly separated. In addition, I wouldn't want to require that both a parameter description and override have to exist, but a user can set either.

Reply:
> The one @hidmic presented as an example implies that there are two different ways to define an override, the original way, and the nested version,

That's correct.

> In my opinion, I think it obfuscates the parameter namespacing syntax by adding even more key words to check.

That sounds a bit like an implementation detail i.e. something a user wouldn't (have to) care about.

In any case, I don't have a strong opinion in favor or against a sibling descriptor mapping or nested descriptors (and neither do others, or at least there's no general consensus), but I do think it should be one of those two collision free approaches and not the current one.